File: cpp/test/utils/metrics/metric_test.cc
#include <string>
#include <gtest/gtest.h>
#include <gmock/gmock.h>

#include "src/utils/metrics/metric.hh"

namespace torchserve {
    class MetricTest : public ::testing::Test {
        protected:
        static const std::string metric_name;
        static const std::vector<Dimension> metric_dimensions;
        static const std::string metric_request_id;
    };

    const std::string MetricTest::metric_name {"test_metric"};
    const std::vector<Dimension> MetricTest::metric_dimensions {
        Dimension("level", "model"),
        Dimension("model_name", "test_model")
    };
    const std::string MetricTest::metric_request_id {"test_request_id"};

    TEST_F(MetricTest, TestMetricValue) {
        MetricValue test_metric_value = MetricValue(1.0, 1664844858, MetricTest::metric_request_id);
        ASSERT_EQ(test_metric_value.GetValue(), 1.0);
        ASSERT_EQ(test_metric_value.GetTimestamp(), 1664844858);
        ASSERT_EQ(test_metric_value.GetRequestId(), MetricTest::metric_request_id);
    }

    TEST_F(MetricTest, TestMetric) {
        Metric test_metric(MetricTest::metric_name, "ms", MetricType::COUNTER, MetricTest::metric_dimensions);
        ASSERT_EQ(test_metric.GetName(), MetricTest::metric_name);
        ASSERT_EQ(test_metric.GetUnit(), "Milliseconds");
        ASSERT_EQ(test_metric.GetType(), MetricType::COUNTER);
        ASSERT_EQ(test_metric.GetDimensions().size(), 2);
        ASSERT_EQ(test_metric.GetDimensions().at(0).GetName(), MetricTest::metric_dimensions.at(0).GetName());
        ASSERT_EQ(test_metric.GetDimensions().at(0).GetValue(), MetricTest::metric_dimensions.at(0).GetValue());
        ASSERT_EQ(test_metric.GetDimensions().at(1).GetName(), MetricTest::metric_dimensions.at(1).GetName());
        ASSERT_EQ(test_metric.GetDimensions().at(1).GetValue(), MetricTest::metric_dimensions.at(1).GetValue());
        ASSERT_TRUE(test_metric.GetValues().empty());
    }

    TEST_F(MetricTest, TestCounterMetric) {
        Metric test_metric(MetricTest::metric_name, "ms", MetricType::COUNTER, MetricTest::metric_dimensions);
        ASSERT_TRUE(test_metric.GetValues().empty());

        test_metric.Update(1.0);
        ASSERT_EQ(test_metric.GetValues().size(), 1);
        ASSERT_EQ(test_metric.GetValues().at(0).GetValue(), 1.0);
        ASSERT_EQ(test_metric.GetValues().at(0).GetRequestId(), "");

        test_metric.Update(2.0, MetricTest::metric_request_id);
        ASSERT_EQ(test_metric.GetValues().size(), 2);
        ASSERT_EQ(test_metric.GetValues().at(0).GetValue(), 1.0);
        ASSERT_EQ(test_metric.GetValues().at(0).GetRequestId(), "");
        ASSERT_EQ(test_metric.GetValues().at(1).GetValue(), 2.0);
        ASSERT_EQ(test_metric.GetValues().at(1).GetRequestId(), MetricTest::metric_request_id);

        test_metric.Update(-1.0);
        ASSERT_EQ(test_metric.GetValues().size(), 2);
        ASSERT_EQ(test_metric.GetValues().at(0).GetValue(), 1.0);
        ASSERT_EQ(test_metric.GetValues().at(0).GetRequestId(), "");
        ASSERT_EQ(test_metric.GetValues().at(1).GetValue(), 2.0);
        ASSERT_EQ(test_metric.GetValues().at(1).GetRequestId(), MetricTest::metric_request_id);

        test_metric.Reset();
        ASSERT_TRUE(test_metric.GetValues().empty());
    }

    TEST_F(MetricTest, TestGaugeMetric) {
        Metric test_metric(MetricTest::metric_name, "count", MetricType::GAUGE, MetricTest::metric_dimensions);
        ASSERT_TRUE(test_metric.GetValues().empty());

        test_metric.Update(1.0);
        ASSERT_EQ(test_metric.GetValues().size(), 1);
        ASSERT_EQ(test_metric.GetValues().at(0).GetValue(), 1.0);
        ASSERT_EQ(test_metric.GetValues().at(0).GetRequestId(), "");

        test_metric.Update(-2.0, MetricTest::metric_request_id);
        ASSERT_EQ(test_metric.GetValues().size(), 2);
        ASSERT_EQ(test_metric.GetValues().at(0).GetValue(), 1.0);
        ASSERT_EQ(test_metric.GetValues().at(0).GetRequestId(), "");
        ASSERT_EQ(test_metric.GetValues().at(1).GetValue(), -2.0);
        ASSERT_EQ(test_metric.GetValues().at(1).GetRequestId(), MetricTest::metric_request_id);

        test_metric.Reset();
        ASSERT_TRUE(test_metric.GetValues().empty());
    }

    TEST_F(MetricTest, TestHistogramMetric) {
        Metric test_metric(MetricTest::metric_name, "count", MetricType::HISTOGRAM, MetricTest::metric_dimensions);
        ASSERT_TRUE(test_metric.GetValues().empty());

        test_metric.Update(1.0);
        ASSERT_EQ(test_metric.GetValues().size(), 1);
        ASSERT_EQ(test_metric.GetValues().at(0).GetValue(), 1.0);
        ASSERT_EQ(test_metric.GetValues().at(0).GetRequestId(), "");

        test_metric.Update(-2.0, MetricTest::metric_request_id);
        ASSERT_EQ(test_metric.GetValues().size(), 2);
        ASSERT_EQ(test_metric.GetValues().at(0).GetValue(), 1.0);
        ASSERT_EQ(test_metric.GetValues().at(0).GetRequestId(), "");
        ASSERT_EQ(test_metric.GetValues().at(1).GetValue(), -2.0);
        ASSERT_EQ(test_metric.GetValues().at(1).GetRequestId(), MetricTest::metric_request_id);

        test_metric.Reset();
        ASSERT_TRUE(test_metric.GetValues().empty());
    }
} // namespace torchserve

File: frontend/server/src/test/java/org/pytorch/serve/metrics/MetricTest.java
package org.pytorch.serve.metrics;

import io.prometheus.client.CollectorRegistry;
import java.io.StringWriter;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.core.Logger;
import org.apache.logging.log4j.core.appender.WriterAppender;
import org.pytorch.serve.metrics.format.prometheous.PrometheusCounter;
import org.pytorch.serve.metrics.format.prometheous.PrometheusGauge;
import org.pytorch.serve.metrics.format.prometheous.PrometheusHistogram;
import org.pytorch.serve.util.ConfigManager;
import org.testng.Assert;
import org.testng.annotations.AfterClass;
import org.testng.annotations.BeforeClass;
import org.testng.annotations.BeforeMethod;
import org.testng.annotations.Test;

@Test(singleThreaded = true)
public class MetricTest {
    private final String testMetricName = "TestMetric";
    private final String testMetricUnit = "ms";
    private final ArrayList<String> testMetricDimensionNames =
            new ArrayList<String>(Arrays.asList("ModelName", "Level", "Hostname"));
    private final ArrayList<String> testMetricDimensionValues =
            new ArrayList<String>(Arrays.asList("TestModel", "Model", "TestHost"));
    private final String testRequestId = "fa8639a8-d3fa-4a25-a80f-24463863fe0f";
    private final Logger loggerModelMetrics =
            (org.apache.logging.log4j.core.Logger)
                    LogManager.getLogger(ConfigManager.MODEL_METRICS_LOGGER);
    private final Logger loggerTsMetrics =
            (org.apache.logging.log4j.core.Logger)
                    LogManager.getLogger(ConfigManager.MODEL_SERVER_METRICS_LOGGER);
    private final String modelMetricsAppenderName = "ModelMetricsAppender";
    private final String tsMetricsAppenderName = "TsMetricsAppender";
    private final StringWriter modelMetricsContent = new StringWriter();
    private final StringWriter tsMetricsContent = new StringWriter();
    private final WriterAppender modelMetricsAppender =
            WriterAppender.createAppender(
                    null, null, modelMetricsContent, modelMetricsAppenderName, true, false);
    private final WriterAppender tsMetricsAppender =
            WriterAppender.createAppender(
                    null, null, tsMetricsContent, tsMetricsAppenderName, true, false);

    @BeforeClass
    public void registerMetricLogAppenders() {
        loggerModelMetrics.addAppender(modelMetricsAppender);
        modelMetricsAppender.start();
        loggerTsMetrics.addAppender(tsMetricsAppender);
        tsMetricsAppender.start();
    }

    @BeforeMethod
    public void flushLogWriterStreams() {
        modelMetricsContent.flush();
        tsMetricsContent.flush();
    }

    @BeforeMethod
    public void clearPrometheusRegistry() {
        CollectorRegistry.defaultRegistry.clear();
    }

    @AfterClass
    public void unregisterMetricLogAppenders() {
        modelMetricsAppender.stop();
        loggerModelMetrics.removeAppender(modelMetricsAppender);
        tsMetricsAppender.stop();
        loggerTsMetrics.removeAppender(tsMetricsAppender);
    }

    @Test
    public void testBackendLogMetric() {
        IMetric testMetric =
                MetricBuilder.build(
                        MetricBuilder.MetricMode.LOG,
                        MetricBuilder.MetricType.COUNTER,
                        testMetricName,
                        testMetricUnit,
                        testMetricDimensionNames);
        Assert.assertEquals(testMetric.getClass(), LogMetric.class);

        testMetric.addOrUpdate(testMetricDimensionValues, testRequestId, 1.0);
        String expectedMetricString =
                "TestMetric.ms:1.0|#ModelName:TestModel,Level:Model|#hostname:TestHost,"
                        + "requestID:fa8639a8-d3fa-4a25-a80f-24463863fe0f,timestamp:";
        Assert.assertTrue(modelMetricsContent.toString().contains(expectedMetricString));
    }

    @Test
    public void testFrontendLogMetric() {
        IMetric testMetric =
                MetricBuilder.build(
                        MetricBuilder.MetricMode.LOG,
                        MetricBuilder.MetricType.GAUGE,
                        testMetricName,
                        testMetricUnit,
                        testMetricDimensionNames);
        Assert.assertEquals(testMetric.getClass(), LogMetric.class);

        testMetric.addOrUpdate(testMetricDimensionValues, 1.0);
        String expectedMetricString =
                "TestMetric.ms:1.0|#ModelName:TestModel,Level:Model|#hostname:TestHost,timestamp:";
        Assert.assertTrue(tsMetricsContent.toString().contains(expectedMetricString));
    }

    @Test
    public void testBackendPrometheusCounter() {
        IMetric testMetric =
                MetricBuilder.build(
                        MetricBuilder.MetricMode.PROMETHEUS,
                        MetricBuilder.MetricType.COUNTER,
                        testMetricName,
                        testMetricUnit,
                        testMetricDimensionNames);
        Assert.assertEquals(testMetric.getClass(), PrometheusCounter.class);

        testMetric.addOrUpdate(testMetricDimensionValues, testRequestId, 1.0);
        Double metricValue =
                CollectorRegistry.defaultRegistry.getSampleValue(
                        testMetricName,
                        testMetricDimensionNames.toArray(new String[0]),
                        testMetricDimensionValues.toArray(new String[0]));
        Assert.assertEquals(metricValue, Double.valueOf(1.0));
        testMetric.addOrUpdate(testMetricDimensionValues, testRequestId, 2.0);
        metricValue =
                CollectorRegistry.defaultRegistry.getSampleValue(
                        testMetricName,
                        testMetricDimensionNames.toArray(new String[0]),
                        testMetricDimensionValues.toArray(new String[0]));
        Assert.assertEquals(metricValue, Double.valueOf(3.0));
    }

    @Test
    public void testFrontendPrometheusCounter() {
        IMetric testMetric =
                MetricBuilder.build(
                        MetricBuilder.MetricMode.PROMETHEUS,
                        MetricBuilder.MetricType.COUNTER,
                        testMetricName,
                        testMetricUnit,
                        testMetricDimensionNames);
        Assert.assertEquals(testMetric.getClass(), PrometheusCounter.class);

        testMetric.addOrUpdate(testMetricDimensionValues, 1.0);
        Double metricValue =
                CollectorRegistry.defaultRegistry.getSampleValue(
                        testMetricName,
                        testMetricDimensionNames.toArray(new String[0]),
                        testMetricDimensionValues.toArray(new String[0]));
        Assert.assertEquals(metricValue, Double.valueOf(1.0));
        testMetric.addOrUpdate(testMetricDimensionValues, 2.0);
        metricValue =
                CollectorRegistry.defaultRegistry.getSampleValue(
                        testMetricName,
                        testMetricDimensionNames.toArray(new String[0]),
                        testMetricDimensionValues.toArray(new String[0]));
        Assert.assertEquals(metricValue, Double.valueOf(3.0));
    }

    @Test
    public void testBackendPrometheusGauge() {
        IMetric testMetric =
                MetricBuilder.build(
                        MetricBuilder.MetricMode.PROMETHEUS,
                        MetricBuilder.MetricType.GAUGE,
                        testMetricName,
                        testMetricUnit,
                        testMetricDimensionNames);
        Assert.assertEquals(testMetric.getClass(), PrometheusGauge.class);

        testMetric.addOrUpdate(testMetricDimensionValues, testRequestId, 1.0);
        Double metricValue =
                CollectorRegistry.defaultRegistry.getSampleValue(
                        testMetricName,
                        testMetricDimensionNames.toArray(new String[0]),
                        testMetricDimensionValues.toArray(new String[0]));
        Assert.assertEquals(metricValue, Double.valueOf(1.0));
        testMetric.addOrUpdate(testMetricDimensionValues, testRequestId, 2.0);
        metricValue =
                CollectorRegistry.defaultRegistry.getSampleValue(
                        testMetricName,
                        testMetricDimensionNames.toArray(new String[0]),
                        testMetricDimensionValues.toArray(new String[0]));
        Assert.assertEquals(metricValue, Double.valueOf(2.0));
    }

    @Test
    public void testFrontendPrometheusGauge() {
        IMetric testMetric =
                MetricBuilder.build(
                        MetricBuilder.MetricMode.PROMETHEUS,
                        MetricBuilder.MetricType.GAUGE,
                        testMetricName,
                        testMetricUnit,
                        testMetricDimensionNames);
        Assert.assertEquals(testMetric.getClass(), PrometheusGauge.class);

        testMetric.addOrUpdate(testMetricDimensionValues, 1.0);
        Double metricValue =
                CollectorRegistry.defaultRegistry.getSampleValue(
                        testMetricName,
                        testMetricDimensionNames.toArray(new String[0]),
                        testMetricDimensionValues.toArray(new String[0]));
        Assert.assertEquals(metricValue, Double.valueOf(1.0));
        testMetric.addOrUpdate(testMetricDimensionValues, 2.0);
        metricValue =
                CollectorRegistry.defaultRegistry.getSampleValue(
                        testMetricName,
                        testMetricDimensionNames.toArray(new String[0]),
                        testMetricDimensionValues.toArray(new String[0]));
        Assert.assertEquals(metricValue, Double.valueOf(2.0));
    }

    @Test
    public void testBackendPrometheusHistogram() {
        IMetric testMetric =
                MetricBuilder.build(
                        MetricBuilder.MetricMode.PROMETHEUS,
                        MetricBuilder.MetricType.HISTOGRAM,
                        testMetricName,
                        testMetricUnit,
                        testMetricDimensionNames);
        Assert.assertEquals(testMetric.getClass(), PrometheusHistogram.class);

        testMetric.addOrUpdate(testMetricDimensionValues, testRequestId, 1.0);
        Double metricValue =
                CollectorRegistry.defaultRegistry.getSampleValue(
                        testMetricName + "_sum",
                        testMetricDimensionNames.toArray(new String[0]),
                        testMetricDimensionValues.toArray(new String[0]));
        Assert.assertEquals(metricValue, Double.valueOf(1.0));
        testMetric.addOrUpdate(testMetricDimensionValues, testRequestId, 2.0);
        metricValue =
                CollectorRegistry.defaultRegistry.getSampleValue(
                        testMetricName + "_sum",
                        testMetricDimensionNames.toArray(new String[0]),
                        testMetricDimensionValues.toArray(new String[0]));
        Assert.assertEquals(metricValue, Double.valueOf(3.0));
    }

    @Test
    public void testFrontendPrometheusHistogram() {
        IMetric testMetric =
                MetricBuilder.build(
                        MetricBuilder.MetricMode.PROMETHEUS,
                        MetricBuilder.MetricType.HISTOGRAM,
                        testMetricName,
                        testMetricUnit,
                        testMetricDimensionNames);
        Assert.assertEquals(testMetric.getClass(), PrometheusHistogram.class);

        testMetric.addOrUpdate(testMetricDimensionValues, 1.0);
        Double metricValue =
                CollectorRegistry.defaultRegistry.getSampleValue(
                        testMetricName + "_sum",
                        testMetricDimensionNames.toArray(new String[0]),
                        testMetricDimensionValues.toArray(new String[0]));
        Assert.assertEquals(metricValue, Double.valueOf(1.0));
        testMetric.addOrUpdate(testMetricDimensionValues, 2.0);
        metricValue =
                CollectorRegistry.defaultRegistry.getSampleValue(
                        testMetricName + "_sum",
                        testMetricDimensionNames.toArray(new String[0]),
                        testMetricDimensionValues.toArray(new String[0]));
        Assert.assertEquals(metricValue, Double.valueOf(3.0));
    }

    @Test
    public void testLegacyMetricMode() {
        ArrayList<String> legacyPrometheusMetrics =
                new ArrayList<String>(
                        Arrays.asList(
                                "ts_inference_requests_total",
                                "ts_inference_latency_microseconds",
                                "ts_queue_latency_microseconds"));

        for (String metricName : legacyPrometheusMetrics) {
            IMetric testMetric =
                    MetricBuilder.build(
                            MetricBuilder.MetricMode.LEGACY,
                            MetricBuilder.MetricType.COUNTER,
                            metricName,
                            testMetricUnit,
                            testMetricDimensionNames);
            Assert.assertEquals(testMetric.getClass(), PrometheusCounter.class);

            testMetric.addOrUpdate(testMetricDimensionValues, 1.0);
            Double metricValue =
                    CollectorRegistry.defaultRegistry.getSampleValue(
                            metricName,
                            testMetricDimensionNames.toArray(new String[0]),
                            testMetricDimensionValues.toArray(new String[0]));
            Assert.assertEquals(metricValue, Double.valueOf(1.0));
            testMetric.addOrUpdate(testMetricDimensionValues, 2.0);
            metricValue =
                    CollectorRegistry.defaultRegistry.getSampleValue(
                            metricName,
                            testMetricDimensionNames.toArray(new String[0]),
                            testMetricDimensionValues.toArray(new String[0]));
            Assert.assertEquals(metricValue, Double.valueOf(3.0));
        }

        // Frontend log metric
        IMetric testMetric =
                MetricBuilder.build(
                        MetricBuilder.MetricMode.LEGACY,
                        MetricBuilder.MetricType.GAUGE,
                        testMetricName,
                        testMetricUnit,
                        testMetricDimensionNames);
        Assert.assertEquals(testMetric.getClass(), LogMetric.class);

        testMetric.addOrUpdate(testMetricDimensionValues, 1.0);
        String expectedMetricString =
                "TestMetric.ms:1.0|#ModelName:TestModel,Level:Model|#hostname:TestHost,timestamp:";
        Assert.assertTrue(tsMetricsContent.toString().contains(expectedMetricString));

        // Backend log metric
        testMetric =
                MetricBuilder.build(
                        MetricBuilder.MetricMode.LEGACY,
                        MetricBuilder.MetricType.COUNTER,
                        testMetricName,
                        testMetricUnit,
                        testMetricDimensionNames);
        Assert.assertEquals(testMetric.getClass(), LogMetric.class);

        testMetric.addOrUpdate(testMetricDimensionValues, testRequestId, 1.0);
        expectedMetricString =
                "TestMetric.ms:1.0|#ModelName:TestModel,Level:Model|#hostname:TestHost,"
                        + "requestID:fa8639a8-d3fa-4a25-a80f-24463863fe0f,timestamp:";
        Assert.assertTrue(modelMetricsContent.toString().contains(expectedMetricString));
    }

    @Test
    public void testParseBackendMetricLogWithoutType() {
        String backendMetricLog =
                "HandlerTime.Milliseconds:71.77|#ModelName:mnist,Level:Model|#hostname:test-host,1699061430,6d1726a7-172c-4010-b671-01d71bacc451";
        Metric parsedMetric = Metric.parse(backendMetricLog);

        Assert.assertEquals("HandlerTime", parsedMetric.getMetricName());
        Assert.assertEquals("Milliseconds", parsedMetric.getUnit());
        Assert.assertEquals("71.77", parsedMetric.getValue());
        List<String> dimensionNames = new ArrayList<String>();
        for (Dimension dimension : parsedMetric.getDimensions()) {
            dimensionNames.add(dimension.getName());
        }
        Assert.assertEquals(Arrays.asList("ModelName", "Level"), dimensionNames);
        List<String> dimensionValues = new ArrayList<String>();
        for (Dimension dimension : parsedMetric.getDimensions()) {
            dimensionValues.add(dimension.getValue());
        }
        Assert.assertEquals(Arrays.asList("mnist", "Model"), dimensionValues);
        Assert.assertEquals(null, parsedMetric.getType());
        Assert.assertEquals("test-host", parsedMetric.getHostName());
        Assert.assertEquals("1699061430", parsedMetric.getTimestamp());
        Assert.assertEquals("6d1726a7-172c-4010-b671-01d71bacc451", parsedMetric.getRequestId());
    }

    @Test
    public void testParseBackendMetricLogWithType() {
        String backendMetricLog =
                "PredictionTime.Milliseconds:71.95|#ModelName:mnist,Level:Model|#type:GAUGE|#hostname:test-host,1699061430,6d1726a7-172c-4010-b671-01d71bacc451";
        Metric parsedMetric = Metric.parse(backendMetricLog);

        Assert.assertEquals("PredictionTime", parsedMetric.getMetricName());
        Assert.assertEquals("Milliseconds", parsedMetric.getUnit());
        Assert.assertEquals("71.95", parsedMetric.getValue());
        List<String> dimensionNames = new ArrayList<String>();
        for (Dimension dimension : parsedMetric.getDimensions()) {
            dimensionNames.add(dimension.getName());
        }
        Assert.assertEquals(Arrays.asList("ModelName", "Level"), dimensionNames);
        List<String> dimensionValues = new ArrayList<String>();
        for (Dimension dimension : parsedMetric.getDimensions()) {
            dimensionValues.add(dimension.getValue());
        }
        Assert.assertEquals(Arrays.asList("mnist", "Model"), dimensionValues);
        Assert.assertEquals("GAUGE", parsedMetric.getType());
        Assert.assertEquals("test-host", parsedMetric.getHostName());
        Assert.assertEquals("1699061430", parsedMetric.getTimestamp());
        Assert.assertEquals("6d1726a7-172c-4010-b671-01d71bacc451", parsedMetric.getRequestId());
    }
}
